whatthelog/prefixtree/evaluator.py

```diff
@@ -0,0 +1,167 @@
+import os
+
+
+from scripts.match_trace import match_trace
+from whatthelog.prefixtree.graph import Graph
+from whatthelog.syntaxtree.syntax_tree import SyntaxTree
+
+
+class Evaluator:
+    """
+    Class containing methods for evaluating state models.
+    """
+
+    def __init__(self,
+                 model: Graph,
+                 syntax_tree: SyntaxTree,
+                 positive_traces_dir: str,
+                 negative_traces_dir: str,
+                 initial_size: int = None):
+
+        self.model = model
+        self.syntax_tree = syntax_tree
+        self.positive_traces_dir = positive_traces_dir
+        self.negative_traces_dir = negative_traces_dir
+        self.initial_model_size = len(model) if initial_size is None else initial_size
+
+    def update(self, new_model: Graph):
+        """
+        Updates the model to test
+        """
+        self.model = new_model
+
+    def evaluate_accuracy(self, debug=False) -> float:
+        """
+        Statically evaluates a model in terms of specificity and recall. The returned
+        Value is the BCR (binary classification rate) defined as `(accuracy + recall) / 2`
+        :param debug: Whether or not to print debug information to the console.
+        """
+        specificity = self.calc_specificity(debug=debug)
+        recall = self.calc_recall(debug=debug)
+
+        return (specificity + recall) / 2
+
+    def evaluate_size(self) -> float:
+        """
+        Evaluates a model in terms of its size. The result is normalized by dividing by the initial model size.
+        """
+        return len(self.model) / self.initial_model_size
+
+    def evaluate(self,
+                 w_accuracy: float = 0.5,
```

Comment:
Maybe the default values for w_accuracy and w_size should be final attributes of the class?

Reply:
That would make calling the `evaluate` function more concise indeed. However, some approaches may want to dynamically change the weights. I'll add support for both.